`Lectures`: Filter lectures on course management page

sunsetcoder opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
#### General
<!-- You only need to choose one of the first two check items: Generally, test on the test servers. -->
<!-- If it's only a small change, testing it locally is acceptable and you may remove the first checkmark. If you are unsure, please test on the test servers. -->
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).

#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->

It allows instructors to filter their lectures based on the start and end dates which can be useful if the course has many lectures and the instructor wants to reduce clutter by filtering out past lectures or quickly find a lecture that is currently in progress to add an attachment. It also allows instructors to find lectures for which the start or end date(s) are undefined.

<!-- If it fixes an open issue, please link to the issue here. -->

### Description
<!-- Describe your changes in detail -->

Added a Filter button with a dropdown menu and 4 options:
- Past: Filters lectures for which the end date is before the current local time.
- Current: Filters lectures for which the start date is same or before the current local time and the end date is after the current time (i.e. lectures currently in progress)
- Future: Filters lectures for which the start date is after the current local time.
- Unspecified Date(s): Filters Lectures for which the start or end date is undefined.

### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->
Prerequisites:
- Instructor

1. Log in to Artemis
2. Navigate to Course Management page
3. Click on Lectures
4. Create lectures with start dates in the past, current, and future, and some without start/end dates specified.
5. Use the filter dropdown menu to ensure that the appropriate lectures are filtered out.

### Review Progress
<!-- Each Pull Request should be reviewed by at least two other developers. The code as well as the functionality (= manual test) needs to be reviewed. -->
<!-- The reviewer or author check the following boxes depending on what was reviewed or tested. All boxes should be checked before merge. -->
<!-- You can add additional checkboxes if it makes sense to only review parts of the code or functionality. -->
<!-- When changes are pushed, uncheck the affected boxes. (Not all changes require full re-reviews.) -->

#### Code Review
- [ ] Review 1
- [ ] Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2

### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->

![image](https://user-images.githubusercontent.com/33063132/171255035-24f6facd-51d2-4a7c-8a32-92937ccded3d.png)
![image](https://user-images.githubusercontent.com/33063132/171255132-6fde2b07-22e2-4c54-89b3-e4cf728e8447.png)

Dominik-Weinzierl commented:
I guess "unspezifiziert Datteln" is the wrong translation ;) 
![image](https://user-images.githubusercontent.com/60005702/170929889-9892c545-fcf6-4fa5-9819-6d8be5775836.png)

ge65cer commented:
Thanks for your contribution. 🎉 Once you check off all items on the checklist, make sure to switch the PR from 'draft' to 'ready for review'.
Also, it would be great if you could add a client test for the `applyFilters` method.

krusche commented:
Please use the same filter features as in other cases. In particular the filter should change the color if active and should indicate how many options are selected.

sunsetcoder commented:
> I guess "unspezifiziert Datteln" is not the intended translation ;) ![image](https://user-images.githubusercontent.com/60005702/170929889-9892c545-fcf6-4fa5-9819-6d8be5775836.png)

Thanks for your feedback. What would be the correct translation for Unspecified Date(s) ? Is "Unbestimmter Zeitpunkt" correct?

sunsetcoder commented:
> Please use the same filter features as in other cases. In particular the filter should change the color if active and should indicate how many options are selected.

Thanks for your feedback. The same filter feature is now implemented as in the other cases. It now changes color if active and indicates how many options are selected.

Also added a Cypress test for the filter feature, but unable to run the test due to no access rights for the Cypress test suite.

Screenshots updated in the original pull request above.

JohannesStoehr commented on src/main/webapp/app/lecture/lecture.component.ts:
```diff
@@ -117,4 +132,37 @@ export class LectureComponent implements OnInit, OnDestroy {
+        let filteredLectures: Array<Lecture> = [];
+
+        // update filteredLectures based on the selected filter option checkboxes
+        const pastLectures = this.lectures.filter((lecture) => lecture.startDate !== undefined && lecture.endDate?.isBefore(dayjs(currentTime)));
```
Why do we have to have a start date, if the end date is before today?

ge65cer replied:
I think the idea is that lectures that do not have a start _and_ end date belong to the _unspecified date_ group.

JohannesStoehr commented on src/main/webapp/app/lecture/lecture.component.html:
```diff
@@ -5,19 +5,84 @@ <h2 id="page-heading">
-                <fa-icon [icon]="faPlus"></fa-icon>
-                <span jhiTranslate="artemisApp.lecture.home.createLabel">Create Lecture</span>
-            </a>
+        <div class="d-flex flex-row-reverse bd-highlight mb-3">
```
Since you're reusing the filter from other pages, would it be wise to move it into a common module and reuse the module in the occurrences?

JohannesStoehr reviewed (commented):
Looked through the code and left two comments:

MetinLamby reviewed (commented):
i tested on ts2 and for me there is no filter button
<img width="1411" alt="Screenshot 2022-07-10 at 12 26 24" src="https://user-images.githubusercontent.com/44767199/178141024-cb768b3a-f95d-4a66-a4ea-424d3245bed9.png">

krusche commented:
This PR is open for a long time now which is bad practice.
@ge65cer could you please take over and help that we merge it in the next few working days? Thank you!

ge65cer commented:
Closed due to merge conflicts. Since this is a fork, I can not push to this branch to resolve them.
